File: candy.cc
// Candy
class Solution {
public:
  int candy(vector<int> &ratings) {
    vector<int> a(ratings.size(), 1);
    for (int i = 1; i < ratings.size(); i++)
      if (ratings[i-1] < ratings[i])
        a[i] = a[i-1]+1;
    for (int i = ratings.size()-2; i >= 0; i--)
      if (ratings[i] > ratings[i+1])
        a[i] = max(a[i], a[i+1]+1);
    return accumulate(a.begin(), a.end(), 0);
  }
};

File: candy.rs
impl Solution {
    pub fn candy(ratings: Vec<i32>) -> i32 {
        let mut a = vec![1; ratings.len()];
        for i in 1..a.len() {
            if ratings[i] > ratings[i-1] { a[i] = a[i-1]+1; }
        }
        for i in (0..a.len()-1).rev() {
            if ratings[i] > ratings[i+1] { a[i] = a[i].max(a[i+1]+1); }
        }
        a.iter().sum()
    }
}
